### PR TITLE
bluesky template format update

### PR DESCRIPTION
**Pull request by rcmckee**
removed brackets from around the text that has links.

### CODE REVIEW

**Comment by mlissner** [comment]
Thanks @rcmckee! Can you sign the CLA above, please?

File: .replit
Comment (mlissner) [inline .replit:1]: Do we need this file? I feel like I'd rather ignore it in .gitignore than include it? I don't know much about replit....

**Review by mlissner — APPROVED** [review]
Looks about right to me, but I can't remember if we need any magic to make templates work automatically, so off to @ERosendo for a real review. :)

**Comment by rcmckee** [comment]
the replit needs to be ignored. I was using replit to work on the code from my phone.
I will put that in the ignore file.

Sent with [Proton Mail](https://proton.me/) secure email.

File: .replit
Comment (rcmckee) [reply .replit:1]: I am putting it in my ignore file. It is created when I open the repository from replit.

**Comment by ERosendo** [comment]
Hey @rcmckee, thanks for the PR! @mlissner is absolutely right. Bluesky templates require a bit more behind-the-scenes magic. I apologize for not providing more context in the issue or as a comment in the code.

Unfortunately, updating just the template won't resolve the issue. Back when we introduced Bluesky support in PR #429 , @mlissner raised an important point in one of his [comments](https://github.com/freelawproject/bigcases2/pull/429#pullrequestreview-1763429490). Bluesky places the responsibility for parsing elements like hashtags, links (both single and embedded), and even social card creation on the client (Our bot) itself. This means Bluesky templates go through an additional processing layer before the actual text reaches the Bluesky API. This layer is very important – it extracts information embedded in the template and cleans it before allowing our bots to post.

Therefore, to resolve the current issue, we may need to refine the template and adjust the processing layer to accurately manage the information.

The code responsible for cleaning and parsing elements from Bluesky templates is located at [channels/utils/connectors/bluesky_api/client.py](https://github.com/freelawproject/bigcases2/blob/5e1c74473fd3bac36781d0c3e82ea2a5c7d9fcd8/bc/channel/utils/connectors/bluesky_api/client.py)

For embedded links, we've chosen a format similar to Markdown syntax. We enclose the link text in brackets [] and the actual URL in parentheses (). This makes the text easy to parse.

To handle this format, we have a helper function called [_parse_embedded_links](https://github.com/freelawproject/bigcases2/blob/5e1c74473fd3bac36781d0c3e82ea2a5c7d9fcd8/bc/channel/utils/connectors/bluesky_api/client.py#L150-L192). This function extracts the text and URL from the template, allowing us to create the embedded link. (For a deeper dive into links, mentions, and rich text formatting, refer to the following documentation:  https://docs.bsky.app/docs/advanced-guides/post-richtext). 

To further refine the content, we use the [_clean_text](https://github.com/freelawproject/bigcases2/blob/5e1c74473fd3bac36781d0c3e82ea2a5c7d9fcd8/bc/channel/utils/connectors/bluesky_api/client.py#L194-L210) helper function. This function removes any links that may have been embedded within the text, resulting in a clean final presentation in our posts.

Here are a few viable changes to fix this isue:

1. I believe it's best to maintain the brackets in the template. They serve as indicators for identifying embedded links. Instead of removing them, we can focus on the `_clean_text` and `_parse_embedded_links` methods. These methods can remove the brackets from the final text while ensuring the proper creation of the data required for displaying the link within Bluesky. 

2. in the `_parse_embedded_links` method, we use a [regular expression](https://github.com/freelawproject/bigcases2/blob/5e1c74473fd3bac36781d0c3e82ea2a5c7d9fcd8/bc/channel/utils/connectors/bluesky_api/client.py#L170-L172) to identify URLs enclosed within parentheses and the corresponding text within brackets. These elements are captured into two distinct named groups: `uri` for the URL and `name` for the text with brackets. By analyzing the starting and ending positions of the `name` group and using the `uri` group to calculate an offset, we can accurately compute the final text's position without including the URL portion. 
   
   We need to update the regular expression to exclude brackets from the `name` group and adjust the offset calculation formula to correctly compute the final text(which should exclude both brackets and the enclosed URL). Here's a suggested approach:

```diff
    def _parse_embedded_links(self, text: str) -> list[RegexMatch]:
        """
        Parses embedded links from text.

        This helper method attempts to identify and return all occurrences of
        link markup in the provided text. If no occurrences are found, an empty
        list is returned.

        Args:
            text (str): The text to parse.

        Returns:
            list[RegexMatch]: List of matches.
        """
        spans = []
+       # Delimiters for embedded link URLs
+       link_url_open = r"[\[]"
+       link_url_close = r"[\]]"
        # Matches anything that isn't a square closing bracket
        name_regex = "[^]]+"
        # Matches http:// or https:// followed by anything but a closing parenthesis
        url_regex = "http[s]?://[^)]+"
        # Combined regex expression with named groups.
        markup_regex = (
-           rf"(?P<name>\[{name_regex}])(?P<uri>\(\s*{url_regex}\s*\))"
+           rf"(?P<opening_tag>{link_url_open})(?P<name>{name_regex})(?P<closing_tag>{link_url_close})(?P<uri>\(\s*{url_regex}\s*\))"
        ).encode()
        text_bytes = text.encode("UTF-8")
        offset = 0
        for m in re.finditer(markup_regex, text_bytes):
            # Remove parenthesis and whitespaces from the uri
            cleaned_uri = re.sub(
                r"\(\s*|\s*\)", "", m.group("uri").decode("UTF-8")
            )
            # We need the offset variable to fine-tune the target word's
            # position and ensure the link is created in the right spot
            # because We run the regex on the full text with links, but
            # only post the cleaned-up version without them
            spans.append(
                RegexMatch(
-                   start=m.start("name") - offset,
-                   end=m.end("name") - offset,
+                   start=m.start("opening_tag") - offset,
+                   end=m.end("name") - len(m.group("opening_tag")) - offset,
                    text=cleaned_uri,
                )
            )
-           offset += len(m.group("uri"))
+           offset += (
+               len(m.group("uri"))
+             + len(m.group("opening_tag"))
+             + len(m.group("closing_tag"))
            )
        return spans
```

3. in the `_clean_text`, we'll need to refine the regular expression to remove the brackets: Here's a suggestion: 

```diff
    def _clean_text(sef, text: str) -> str:
        """
        Removes all link markup notations from a given text, leaving only the
        plain text content.

        This helper function specifically targets the markup used for embedding
        hyperlinks within the provided text. It aims to strip away all URLs
        associated with links, leaving behind the raw textual information.

        Args:
            text (str): the text to be cleansed of link markup.

        Returns:
            str: string containing the original text with all link markup
            notations removed.
        """
-       return re.sub(r"(?<=])\(\S+\)", "", text)
+       return re.sub(r"(?<=])\(\S+\)|[\[]|[\]]", "", text)

```

Thanks for your help! Let me know if you have any questions.

**Comment by rcmckee** [comment]
Ok I got it.

It seemed a little too simple to work.

I will tackle this issue.

Best,
Robert

**Comment by mlissner** [comment]
@rcmckee, if you're tinkering with these templates, I wonder if this issue could fit into this PR too: https://github.com/freelawproject/bigcases2/issues/596

It'd be nice to get this fixed and I think it shouldn't be much more trouble? No worries if not!

**Comment by rcmckee** [comment]
I am closing this pull request to generate a new one with the suggested changes from https://github.com/freelawproject/bigcases2/pull/583#issuecomment-2327070111

**Pull request closed without merging**